Experimental support for OPi5 Pro

gnull opened the pull request:
This PR brings experimental system and sd image configs for OPi5 Pro. The configs are based on OPi5 and OPi5 Plus, and are untested since I don't own a Pro board. Looking for feedback from someone who can run try and run it.

## Running instructions

Build with `nix build .#sdImage-opi5pro` or `nix build .#sdImage-opi5pro-cross` (if you're cross compiling from non-ARM computer).

Before running the image, you should run [this](https://www.armbian.com/orange-pi-5-pro/) Armbian image and follow step 1 from [here](https://github.com/gnull/nixos-rk3588/blob/main/U-Boot.md) to install the bootloader into SPI.

## Current State

The current config fully repeats that of OPi5 Plus, only changing the DTS used. Whether something else needs to change is to be determined.

gnull committed:
Add an experimental OPi5Pro image

Redhawk18 commented:
What would you recommend for the boot loader? My general plan is this right now

1. dd if=edk-orangepi5pro of=/dev/sdb
2. create root partition for nixos
3. dd if=orangepi5pro-nixos.img of=/dev/sdb2

Redhawk18 commented:
I ordered and got my emmc today and its not detected in armbian of course.

Redhawk18 commented:
update: official orangepi ubuntu os shows emmc, so it might be a bug in armbian/dietpi...

Redhawk18 commented:
update: installed  boot loader through orangepi os, armbian and dietpi still are bugged but require so much reproducibility information I don't have the time to reburn their images and fill out the forms.

Redhawk18 commented:
update:

> <<< Welcome to NixOS 24.11.20241018.4c2fcb0 (aarch64) -tty1 >>>

😍

Redhawk18 commented:
so to confirm, nixos is installed. I just can't boot it... I have my drive set as a swapfile with a root ext4 partition with a boot folder.

gnull commented:
Cool. At what point does the boot fail right now? Could you post the exact log or a picture of what you see?

Redhawk18 commented:
The uuids were correct, it did give me a shell. I think I bricked my emmc aswell.

Redhawk18 commented:
So whats the proper way to get this working then? Do I still need uboot from orangepi or armbian? I wish it was as simple as plugging in the sdcard and formatting the ssh and doing `nixos-install`

gnull commented:
If you haven't changed the UUID in kernel parameters and UUID in the image's root partition, it should be fine. You can check the UUID of the image you generated with Gparted, for example. Are you directly using the image you got from `nix build`?

If you have emmc, flash Armbian U-Boot to it (with `armbian-install` if it works), then flash the image you get from `nix build` to your drive (SD card or NVMe). This was the ideal scenario I had in mind. If this boots, then it's easy to customize.

What I wrote before about `dd`-ing the U-Boot to SD card was for the case when you don't have emmc. There's more uncertainty here, so if emmc is there, try what I described in the previous paragraph.

gnull commented:
We need to boot something first, we can talk about `nixos-install` only after some NixOS is booted.

[This](https://github.com/gnull/nixos-rk3588/pull/8#issuecomment-3685723830) comment is what I think can work without emmc. You need 3 binaries:

1. disk image you built with `nix build` from this repo (it should be U-Boot image version, not UEFI),
2. idbloader,
3. u-boot.

The 2-3 you can find somewhere in Armbian. Maybe you can extract them from Armbian's own rootfs, they usually come with u-boot `.deb` package installed that contains the binaries which `armbian-install` writes into emmc/SD when you tell it to install u-boot. We want those binaries. If you open the root partition of [Minimal image](https://www.armbian.com/orange-pi-5-pro/#kernels-archive-all), there's `u-boot.itb` and `idbloader.img` under `/usr/lib/linux-u-boot-vendor-orangepi5pro/`, these look like they should be 1 and 2 from my list above.

 You `dd` the 1-2 under the right offsets into 1 ([this](https://salsa.debian.org/debian/u-boot/-/blob/debian/latest/debian/u-boot-rockchip.README.Debian) looks like the right dd commands, put your image 1 into `of=` parameter, and 1-2 into `if=`).

Then `dd` the result into your SD card, check with Gparted that it looks 'alright' (boot and rootfs partitions are there, UUID looks fine) and try to boot.

I haven't followed these steps myself, just gave it a quick investigation, something may need to be adjusted after you try.

AleAndForCode commented:
I successfully booted my opi5pro using this image. Here's what worked:

1. Extracted u-boot.itb and idbloader.img from Armbian and flashed them to a TF card
2. Flashed the NixOS image to a USB stick (temporary setup while waiting for NVMe SSD)
3. The opi5pro can only boot from USB 2.0 ports according to the manual. Boot time with this setup is about 5 minutes...

I'm having issues with nixos-rebuild. Generated default configs, added a user, and rebuilt successfully, but after reboot the system always starts with the initial generation (no boot menu appears). Then I noticed the BOOT partition wasn't mounted. After mounting it and updating configuration, I got "no space left" error during rebuild (seems to be copying a new upstream kernel).
I may need to pin the Armbian kernel in my configuration somehow.

So, should the BOOT partition be mounted at /boot? Are there guidelines for establishing a proper configuration.nix/rebuild workflow after initial boot?

Thanks for this work!

gnull committed:
Add prebuilt u-boot binaries for OPi5 Pro, include them in the image

The u-boot Debian package we take binaries from is this:
  https://github.com/armbian/os/pkgs/container/os%2Fuboot-orangepi5pro-vendor

Downloaded via:
  oras pull ghcr.io/armbian/os/uboot-orangepi5pro-vendor:2017.09-sb04e-p9292-h697e-v14ee-b4d1d-r448a

gnull commented:
Glad to hear you got it to boot!

I never tested the Colmena setup, it's in the docs since Ryan's original repo. Maybe I should rewrite the docs to suggest using `nixosConfiguration` like in your example, I'm using it too.

Does it take 3 minutes only on the first boot, or each time? The first boot could be less surprising if it's resizing the partitions to take the whole SD card (Armbian does that, so maybe NixOS too).

https://github.com/gnull/nixos-rk3588/blob/31e2ad4e722b8eaad9de2bdac37b3619daa535e3/modules/sd-image/rock5a.nix#L57

Maybe I'll find a good way to automatically install u-boot in the image we produce, automating the `dd` steps in the flake, then we can provide something that just works. It's easiest to probably just take the binaries from Armbian image and commit them in the repo here like we do for 5A.

Are you looking for boot menu over HDMI or UART? If it's UART, it could be baud rate or some other connection setting (sometimes, u-boot and kernel use different settings so if you launch minicom with kernel settings, you won't see u-boot). In any case, u-boot environment is a relevant keyword, you can try to see if there's any connection settings specified in the env file. Or look up what UART settings Armbian suggests and try them (since we use their u-boot).

gnull commented:
I pushed a change that should correctly embed the u-boot binaries in the produced image. Now, a simple `nix build .#sdImage-opi5pro` should make a runnable SD card image, no do `dd` u-boot.

Let me know if it works (since I can't test it myself).

gnull commented:
https://github.com/NixOS/nixpkgs/blob/9f0c42f8bc7151b8e7e5840fb3bd454ad850d8c5/pkgs/misc/uboot/default.nix#L38

Upstream u-boot doesn't have a config file for 5 Pro, it's probably not supported yet. Looks like vendor binaries may be our best option until they do.
